### PR TITLE
feat: clean (obfuscate/redact) specs in memory

### PR DESCRIPTION
**Pull request by xiangce**
- and add more tests
- see RHINENG-8775

### All Pull Requests:

Check all that apply:

* [x] Have you followed the guidelines in our Contributing document, including the instructions about commit messages?
* [ ] Is this PR to correct an issue?
* [x] Is this PR an enhancement?

### Complete Description of Additions/Changes:

*Add your description here*

### CODE REVIEW

File: insights/core/spec_cleaner.py [hunk 1]
```diff
@@ -336,35 +345,54 @@ def _redact_line(self, line):
...
         for line in lines:
-            # 1. Do Redaction by default, unless 'no_redact=True'
-            if line and not no_redact:
-                line = self._redact_line(line)
-            # 2. TODO: Do Filtering as per the "filters"
-            # line = self._filter_line_allow(line, filters)
-            # 3. Do Obfuscation as per the "obf_funcs"
-            line = self._obfuscate_line(line, obf_funcs or [])
+            line = _clean_line(line)
             result.append(line) if line is not None else None
```
Comment (JoySnow) [inline insights/core/spec_cleaner.py:396]: This `if line is not None` checking could introduce an interesting corner case: 
   - For a collected content, after cleaner.clean_content(), it has the
`result`  `['\n', '\n']`. 

Should this be collected as the final content of the file?     
Or this file should be emptied and skipped in insights-archive?  

See the test-case example in https://github.com/RedHatInsights/insights-core/pull/4061/files .

**Review by JoySnow — CHANGES_REQUESTED** [review]
Hi @xiangce , the changes LGTM, except the two comments below.  Please kindly let me know your thoughts.

File: insights/core/spec_factory.py [hunk 2]
```diff
@@ -71,6 +70,36 @@ def stream(self):
     def _stream(self):
         raise NotImplementedError()
 
+    def _clean_content(self, content):
```
[the remark this answers is not in the document]
Comment (xiangce) [reply insights/core/spec_factory.py:73]: Good catching. I have applied different versions of implementation before submitting the final version. One of them is to do the "cleaning" while `load()` the content, which means, in the `memory` during the collection launched by the insights-client, the `content` will be obfuscated. However, the disadvantage is, for some datasources, it's necessary to get the original IP/FQDN to collect the required data. Then I turned to do the `cleaning` when writing the content to the disk. but forget to change the `_clean_content` back to work with the `self.conent` directly. 😿

**Review by JoySnow — APPROVED** [review]
@xiangce  The changes LGTM now.

File: insights/core/spec_cleaner.py (hunk 1 above)
Comment (xiangce) [reply insights/core/spec_cleaner.py:396]: Per our discussion in yesterday's sync, I opened a new card to track the excluding of the empty files RHINENG-8978

File: insights/core/spec_factory.py [hunk 3]
```diff
@@ -71,6 +70,30 @@ def stream(self):
     def _stream(self):
         raise NotImplementedError()
 
+    def _clean_content(self):
+        """
+        Clean (Obfuscate, Redact, and Filter) the Spec Content ONLY when doing
```
Comment (JoySnow) [inline insights/core/spec_factory.py:75]: This `Filter` in docstrings should also be removed according to the code changes.

File: insights/core/spec_factory.py [hunk 4]
```diff
@@ -268,24 +270,11 @@ class TextFileProvider(FileProvider):
...
         self.loaded = True
-        args = self.create_args()
-        if args:
-            rc, out = self.ctx.shell_out(args, keep_rc=True, env=SAFE_ENV)
-            self.rc = rc
-            return out
 
         with safe_open(self.path, "r", encoding="utf-8", errors="surrogateescape") as f:
-            return [l.rstrip("\n") for l in f]
+            return [l.rstrip("\n") for l in apply_filters(self.ds, f)]
```
Comment (JoySnow) [inline insights/core/spec_factory.py:277]: Seeing that we are going to use `apply_filters()` to replace the `["grep", "-F", filters, self.path]` in both `TextFileProvider` and `CommandOutputProvider`. 

Just the docstrings of  func `apply_filters` mentions that: 
> Applys filters to the lines of a datasource. This function is used only in
    integration tests. Filters are applied in an equivalent but more performant
    way at run time.

So I am wondering if there could some efficiency loss.  

Any reason for this change?

**Review by JoySnow — CHANGES_REQUESTED** [review]

File: insights/core/spec_factory.py (hunk 4 above)
Comment (JoySnow) [reply insights/core/spec_factory.py:277]: Discussed with @xiangce , this filter method related change was mentioned in RHINENG-7854  . Then this LGTM.

Comment (xiangce) [reply insights/core/spec_factory.py]: I tested the `grep -F` will be faster than `apply_filters` line by line. so I reverted it in the latest commit. Thanks.

File: insights/core/spec_factory.py (hunk 3 above)
Comment (xiangce) [reply insights/core/spec_factory.py]: Done

**Review by JoySnow — APPROVED** [review]
LGTM

**Pull request merged**